### PR TITLE
refactor data loader

### PR DESCRIPTION
**Pull request by giovp**
This PR is a refactoring of the dataloader. It introduces the following improvements:
- [x] it is possible to create a dataloader from collections of images/regions in the spatialdata object (before, it was a single image-region pair)
- [x] It performs more thorough validation and preprocess as much as possible.
- [x] it simplifies the `__getitem__` method to minimize the amount of computation that is done at that stage, and leverage pre-computed (transformed) coordinates of the tiles at init stage
- [x] it allows the user to both use the rasterize and the bounding_box query API, By default, the latter is used.
- [x] it allows to generate tiles from circles, polygons or multipolygons. For polygons and multipolygons, the length of the polygon is used to get the extent of the tile.
- [x] It returns either a spatialdata object with the image tile and the associated table, or a pair of image/annotation (either from adata.obs or adata.X).

**Considerations for future work**
While I tried to also make it usable for the labels as regions, I realized early on that for that we might want a drastically different type of pipeline in that case. Specifically, we might not want to iterate over centroids but over chunks.

### CODE REVIEW

**Comment by giovp** [comment]
took me very long to fix that dumb sphinx error re import of tiledataset when torch not available.

decide to use this instead of `contextlib` as before. 

```python
try:
    from spatialdata.dataloader.datasets import ImageTilesDataset
except ImportError:
    ImageTilesDataset = None  # type: ignore[assignment, misc]
``` 

do you think we should also raise a warning/info? a bit against because it'd be always raised at import of spatialdata

**Comment by LucaMarconato** [comment]
> do you think we should also raise a warning/info? a bit against because it'd be always raised at import of spatialdata

I am also against a warning/info because it would always be shown. Maybe overkill, instead of `ImageTilesDataset = None` we could have something like this:
```python
# spatialdata/_utils.py
class ErrorCallable:
	def __init__(self, error_message):
		self.error_message = error_message
	def __call__(self, **kwargs):
		raise ImportError(self.error_message)
```

```
try:
    from spatialdata.dataloader.datasets import ImageTilesDataset
except ImportError:
    from spatialdata._utils import ErrorCallable
    ImageTilesDataset = ErrorCallable('Please install `torch` to make ImageTilesDataset available')  # type: ignore[assignment, misc]

File: src/spatialdata/dataloader/datasets.py [hunk 1]
```diff
@@ -1,188 +1,395 @@
...
         regions_to_images
-            A dictionary mapping the regions element key we want to extract the tiles around to the images element key
-            we want to get the image data from.
+            A mapping betwen region and images. The regions are used to compute the tile centers, while the images are
+            used to get the pixel values.
+        regions_to_coordinate_systems
+            A mapping between regions and coordinate systems. The coordinate systems are used to transform both
+            regions coordinates for tiles as well as images.
+        tile_scale
+            The scale of the tiles. This is used only if the `regions` are `shapes`.
```
[the remark this answers is not in the document]
Comment (giovp) [reply src/spatialdata/dataloader/datasets.py:73]: this dataloader doesn't do labels, only shapes

File: src/spatialdata/dataloader/datasets.py [hunk 2]
```diff
@@ -1,188 +1,395 @@
...
+            If `tile_dim_in_units` is passed, `tile_scale` is ignored.
         tile_dim_in_units
             The dimension of the requested tile in the units of the target coordinate system. This specifies the extent
             of the image each tile is querying. This is not related he size in pixel of each returned tile.
-        tile_dim_in_pixels
-            The dimension of the requested tile in pixels. This specifies the size of the output tiles that we will get,
-             independently of which extent of the image the tile is covering.
-        target_coordinate_system
-            The coordinate system in which the tile_dim_in_units is specified.
+        raster
```
[the remark this answers is not in the document]
Comment (giovp) [reply src/spatialdata/dataloader/datasets.py]: done!

File: src/spatialdata/dataloader/datasets.py [hunk 3]
```diff
@@ -1,188 +1,395 @@
...
+            If True, the images are rasterized using :func:`spatialdata.rasterize`.
+            If False, they are rasterized using :func:`spatialdata.bounding_box_query`.
+        return_annot
```
[the remark this answers is not in the document]
Comment (giovp) [reply src/spatialdata/dataloader/datasets.py:87]: done

File: src/spatialdata/dataloader/datasets.py [hunk 4]
```diff
@@ -1,188 +1,395 @@
...
+            If not None, a value from the table is returned together with the image tile.
+            Only columns in :attr:`anndata.AnnData.obs` and :attr:`anndata.AnnData.X`
+            can be returned. If None, it will return a spatialdata object with only the tuple
+            containing the image and the table value.
```
[the remark this answers is not in the document]
Comment (giovp) [reply src/spatialdata/dataloader/datasets.py:91]: I took a look at that function and not sure would fit well what I'm doing there, as in it's passing views of the annotations instead of copies of the full tables. I think I'd have to modify it in order to use it in that context. I understand that current implementation is also not great though

File: src/spatialdata/dataloader/datasets.py [hunk 5]
```diff
@@ -1,188 +1,395 @@
...
+        raster_kwargs
```
[the remark this answers is not in the document]
Comment (giovp) [reply src/spatialdata/dataloader/datasets.py:92]: added the comment in the docstring. The docstring though redirects to the rasterize function so that the arguments there can be inspected

File: src/spatialdata/dataloader/datasets.py [hunk 6]
```diff
@@ -1,188 +1,395 @@
...
+            Keyword arguments passed to :func:`spatialdata.rasterize` if `raster` is True.
+
+        Returns
+        -------
+        :class:`torch.utils.data.Dataset` for loading tiles from a :class:`spatialdata.SpatialData`.
         """
```
[the remark this answers is not in the document]
Comment (giovp) [reply src/spatialdata/dataloader/datasets.py:98]: done.

File: src/spatialdata/dataloader/datasets.py [hunk 7]
```diff
@@ -1,188 +1,395 @@
...
+        """Validate input parameters."""
+        self._region_key = sdata.table.uns[TableModel.ATTRS_KEY][TableModel.REGION_KEY_KEY]
+        self._instance_key = sdata.table.uns[TableModel.ATTRS_KEY][TableModel.INSTANCE_KEY]
+        available_regions = sdata.table.obs[self._region_key].cat.categories
+        cs_region_image = []  # list of tuples (coordinate_system, region, image)
+
+        # check unique matching between regions and images and coordinate systems
+        assert len(set(regions_to_images.values())) == len(
+            regions_to_images.keys()
+        ), "One region cannot be paired to multiple regions."
```
[the remark this answers is not in the document]
Comment (giovp) [reply src/spatialdata/dataloader/datasets.py:126]: done locally

File: src/spatialdata/dataloader/datasets.py [hunk 8]
```diff
@@ -1,188 +1,395 @@
...
+
+            # check that the elements are supported
+            if get_model(region_elem) in [Labels2DModel, Labels3DModel]:
+                raise NotImplementedError("labels elements are not implemented yet.")
+            if get_model(region_elem) not in [ShapesModel]:
+                raise ValueError("`regions_element` must be a shapes element.")
+            if get_model(image_elem) not in [Image2DModel, Image3DModel]:
+                raise ValueError("`images_element` must be an image element.")
+            if isinstance(image_elem, MultiscaleSpatialImage):
+                raise NotImplementedError("Multiscale images are not implemented yet.")
```
[the remark this answers is not in the document]
Comment (giovp) [reply src/spatialdata/dataloader/datasets.py:144]: I agree on both 1 and 2 and I think when working on the models I'll remove spatial image and multiscale spatial image entirely. I would for now keep it like this and modify accordingly once the models have been reworked

File: src/spatialdata/dataloader/datasets.py [hunk 9]
```diff
@@ -1,188 +1,395 @@
...
+                if isinstance(region_trans, BaseTransformation) and isinstance(image_trans, BaseTransformation):
+                    cs_region_image.append((cs, region_key, image_key))
+            except KeyError as e:
+                raise KeyError(f"region {region_key} not found in `regions_to_coordinate_systems`") from e
+
+        self.regions = list(regions_to_coordinate_systems.keys())  # all regions for the dataloader
+        self.sdata = sdata
+        self.dataset_table = self.sdata.table[
+            self.sdata.table.obs[self._region_key].isin(self.regions)
+        ]  # filtered table for the data loader
```
[the remark this answers is not in the document]
Comment (giovp) [reply src/spatialdata/dataloader/datasets.py:163]: unfortunately it only works for one element and not multiple ones. Also I don't need to reorder

**Comment by giovp** [comment]
>  I wonder if some of the table/shapes manipulation operations should be added as new public functions in that file. Feel free to do it or discard this and just tick this item.

I replied there I don't think they are useful for this purpose.

> Please update the [deep learning example notebook](https://github.com/scverse/spatialdata-notebooks/blob/main/notebooks/examples/densenet.ipynb?rgh-link-date=2023-06-29T17%3A58%3A51Z) with the new APIs. Do you have the Xenium + Visium aligned data? I just realized that I forgot to make the CI upload it to S3, gonna fix it tomorrow/Sat, so in that case you will be able to download it from there. I would merge this PR and update the notebook at the same time, otherwise users will find it broken.

good point, I will take a look at the S3 and contact you on zulip, will do it this weekend

File: src/spatialdata/dataloader/datasets.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (LucaMarconato) [reply src/spatialdata/dataloader/datasets.py:91]: You are right. I'd keep the current implementation then since it's still quite simple, we can always change things later if needed.

File: src/spatialdata/dataloader/datasets.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (LucaMarconato) [reply src/spatialdata/dataloader/datasets.py:102]: Thanks! I'll make a commit with a new proposed docstring to cover also the case in which the return type of `ImageTileDataset` is a `SpatialData` object, please then check.

[the remark this answers is not in the document]
Comment (LucaMarconato) [reply src/spatialdata/dataloader/datasets.py:105]: added

**Comment by LucaMarconato** [comment]
Thanks for implementing the changes! I will send you the link on S3 in Zulip later today.

**Comment by LucaMarconato** [comment]
- [ ] I just noticed that the `ImageTilesDataset` class does not appear in the doc, this should also be fixed please.

File: docs/api.md [hunk 10]
```diff
@@ -30,11 +30,10 @@ Operations on `SpatialData` objects.
     match_table_to_element
     concatenate
     rasterize
-    transform
```
Comment (LucaMarconato) [inline docs/api.md:33]: restoring this, since it's imported from `spatialdata._core.operations`

**Comment by giovp** [comment]
@LucaMarconato can we merge this>?
